### PR TITLE
fix: indexer fixed by paginating over the APIs

### PR DESCRIPTION
**Pull request by manangouhari**
**What does this PR do?**
Fixes the indexer.

**How should this be manually tested?**
By going to the `/tezos/marketplace` page, and opening up pages of individual Cryptobots. Also, try looking at your profile page.

**Additional Context:** I'm not sure if this is the best way to go about it, because this induces a lot of latency.

### CODE REVIEW

File: src/utils/indexer.js [hunk 1]
```diff
@@ -1,5 +1,7 @@
 import { APP_NAME, INDEXER_NETWORK, CONTRACT_ADDRESS } from 'src/defaults';
-import { bytes2Char } from "@taquito/tzip16";
+import { bytes2Char } from '@taquito/tzip16';
+import { OctahedronBufferGeometry } from 'three';
```
Comment (bhaskarSingh) [inline src/utils/indexer.js:3]: Remove

```
import { OctahedronBufferGeometry } from 'three';
import { isJSDocUnknownType } from 'typescript';
```

It's not being used any where.

File: src/utils/indexer.js [hunk 2]
```diff
@@ -181,63 +201,68 @@ export const getXTZPriceInUSD = (usd, mutez) => {
...
+        tokenId: elm.tokenId,
+        uri: nft.uri,
+        isForSale: offer ? offer.isForSale : false,
+        saleValueInMutez: offer ? offer.saleValueInMutez : null,
+        seller: offer ? offer.seller : null,
+        offerDate: offer ? offer.timestamp : null,
+        owner: elm.address,
+      };
+    });
+    return filtered;
```
Comment (bhaskarSingh) [inline src/utils/indexer.js:262]: There are two returns here that seem confusing. Please elaborate on what's happening here.

File: src/utils/indexer.js [hunk 3]
```diff
@@ -248,18 +273,25 @@ const getAllTokenHolders = async () => {
...
+    const num_keys = ledgerMetadataJSON.active_keys;
+    const all_holders = [];
+    let tk;
+    for (let i = 0; i < parseInt(num_keys / 10) + 1; i++) {
+      tk = await fetch(
+        `https://api.better-call.dev/v1/bigmap/${INDEXER_NETWORK}/${
+          tokens.value
+        }/keys?offset=${10 * i}`,
+      );
+      all_holders.push(...(await tk.json()));
```
Comment (bhaskarSingh) [inline src/utils/indexer.js:289]:
```
for (let i = 0; i < parseInt(num_keys / 10) + 1; i++) {
      tk = await fetch(
        `https://api.better-call.dev/v1/bigmap/${INDEXER_NETWORK}/${
          tokens.value
        }/keys?offset=${10 * i}`,
      );
      all_holders.push(...(await tk.json()));
```

Please elaborate on what's going here. I think all fetches are being stacked here, right? Which would load all the bots eventually.

Comment (manangouhari) [reply src/utils/indexer.js:289]: Yep, spot on. 

We find out how many are the total keys, and then we know that page size is 10. 
So, we loop until we get all out bots, on every loop we fetch 10 bots with the correct offset.

File: src/utils/indexer.js (hunk 2 above)
Comment (manangouhari) [reply src/utils/indexer.js:262]: The first return is the return inside `tokenHolder.map`.

The second return is what the function is returning.

Comment (bhaskarSingh) [reply src/utils/indexer.js:260]: Got it.

**Review by bhaskarSingh — APPROVED** [review]
The load time will increase as more NFTs are minted. For now, it's okay. It's a good problem to solve when we encounter it.

**Pull request merged**